inc/xmath.h

```diff
@@ -1449,116 +1471,214 @@ struct quat : public vec<4, QS>
+ * @tparam     S     { description }
+ */
+template <size_t Z, size_t X, size_t U, typename S = XMTYPE>
+struct kalman
+{
+	struct
 	{
-		auto l  = sphere_o - ray_o;
-		auto s  = l.dot(ray_d);
-		auto l2 = l.dot(l);
-		auto r2 = sphere_r * sphere_r;
+		mat<X, 1, S> state;
+		mat<X, X, S> covariance = mat<X, X, S>::I();
+		mat<Z, 1, S> measurement_residual;
+	} estimated;
 
-		if (s < 0 && l2 > r2) { return NAN; }
+	mat<X, X, S> I_xx = mat<X, X, S>::I();
 
-		auto m2 = l2 - (s * s);
+	kalman(vec<X, S>& state = {}) { estimated.state = {state}; }
 
-		if (m2 > r2) { return NAN; }
+	/**
+	 * @brief      { function_description }
+	 *
+	 * @param[in]  state_transition  The state transition
+	 * @param[in]  control_to_state  The control to state
+	 * @param[in]  control           The control
+	 */
+	void time_update(const mat<X, X, S>& state_transition,
+	                 const mat<X, U, S>& control_to_state,
+	                 const mat<U, 1, S>& control,
+	                 const mat<X, X, S>& process_noise_covariance = {})
+	{
+		// Advance state in time by state_transition matrix. Apply control
+		// vector transformed into state space to state.
+		estimated.state = state_transition * estimated.state;
+		estimated.state += control_to_state * control;
 
-		auto q = sqrt(r2 - m2);
+		const auto& F = state_transition;
+		auto&       P = estimated.covariance;
+
+		// Inflate covariance. TODO: study
+		P = F * P * F.transpose() + process_noise_covariance;
+	}
 
-		auto b = l2 > r2;
-		return b * (s - q) + !b * (s + q);
+	/**
+	 * @brief      { function_description }
+	 *
+	 * @param[in]  state_transition              The state transition
+	 * @param[in]  state_to_measurement          The state to measurement
+	 * @param[in]  measurement_noise_covariance  The measurement noise
+	 * covariance
+	 * @param[in]  measurement                   The measurement
+	 */
+	void measurement_update(
+	    const mat<X, X, S>& state_transition,
+	    const mat<Z, X, S>& state_to_measurement,
+	    const mat<Z, 1, S>& measurement,
+	    const mat<Z, Z, S>& measurement_noise_covariance = {})
+	{
+		const mat<X, X, S>& F     = state_transition;
+		const mat<Z, X, S>& H     = state_to_measurement;
+		const mat<X, Z, S>  H_T   = state_to_measurement.transpose();
+		auto&               P     = estimated.covariance;
+		auto&               x_hat = estimated.state;
+
+		// Compute error between most recent measurement, and current estimated
+		// state transformed into measurement domain
+
+		auto meas_res_pre = measurement - state_to_measurement * x_hat;
+
+		// TODO: study
+		mat<Z, Z, S> covariance_res_pre =
+		    H * P * H_T + measurement_noise_covariance;
+
+		// TODO: study
+		mat<X, Z, S> K = P * H_T * covariance_res_pre.invert();
+
+		// TODO: study
+		// Update state estimate: X x 1
+		estimated.state += K * meas_res_pre;
+		// TODO: study
+		// Update state covariance estimate: X x X
+		estimated.covariance = (I_xx - K * H) * estimated.covariance;
```

Comment:
This works, but it'll lead to instabilities because of the matrix inverse when the filter gets too confident. The Joseph update solves this well, cf. step 7 here: https://nyxspace.com/MathSpec/orbit_determination/kalman/#measurement-update .

Reply:
Hmm, but sadly this will require automatic differentiation. I'm looking to use this on micros so I'm not sure how performant that would be.

Reply:
Wait, why does it require automatic differentiation? Step 7 only has an additional matrix inverse (which is gonna be annoying depending on the state size for sure).

Reply:
Because step 7 depends on H tilde which looks to be computed using automatic differentiation, right?

Reply:
Ah, no, that's just the notation I use. It's the sensitivity matrix which maps the observation back into state space. Although you _can_ use autodiff for it, you don't need to because that formulation is always the same.

Reply:
Got it, so in my case H~ is `state_to_measurement`?

Reply:
Yes